components/scream/src/physics/shoc/shoc_length_impl.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef SHOC_LENGTH_IMPL_HPP
+#define SHOC_LENGTH_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::shoc_length(
+  const MemberType&            team,
+  const Int&                   nlev,
+  const Int&                   nlevi,
+  const Scalar&                host_dx,
+  const Scalar&                host_dy,
+  const Scalar&                pblh,
+  const uview_1d<const Spack>& tke,
+  const uview_1d<const Spack>& zt_grid,
+  const uview_1d<const Spack>& zi_grid,
+  const uview_1d<const Spack>& dz_zt,
+  const uview_1d<const Spack>& dz_zi,
+  const uview_1d<const Spack>& wthv_sec,
+  const uview_1d<const Spack>& thetal,
+  const uview_1d<const Spack>& thv,
+  const uview_1d<Spack>&       thv_zi,
+  const uview_1d<Spack>&       brunt,
+  const uview_1d<Spack>&       shoc_mix)
+{
```

Comment:
I would like input on my use of `team_barrier()`. Basically, any function that has a `Kokkos::parallel_*(PerThread)` call, and that sets a value which will be used later in this functions, I call `team_barrier()` after. Is this correct?

Reply:
I am the least experienced in these matters of the three of us, but it seems to me that you only need a barrier between A and B if values computed by a thread in A are needed by another thread in B. Of course, this treats the team dispatch mechanism as a black box, so I'm interested to hear what Jim and Andrew have to say.

Reply:
What Jeff wrote is correct. If you have two loops over the same indices and index k does not depend on k+i for i /= 0, then the two loops don't need a team_barrier between them b/c the index-thread map is identical in the two loops.

A reduction has an implicit team_barrier since each thread gets its own copy of the reduction value, so it too does not need a team_barrier after it.

This is the type of question for which writing a single-file sandbox code can be useful: to play around with scenarios and discover nuances.